src/Slick/Pandoc.hs

```diff
@@ -50,76 +40,99 @@ unPandocM p = do
+
 -- | Load in a source document using the given 'PandocReader', then render the 'Pandoc'
--- into text using the given 'PandocWriter'.
--- Returns a 'Value' wherein the rendered text is set to the "content" key and 
--- any metadata is set to its respective key in the 'Value'
-loadUsing :: PandocReader textType -> PandocWriter -> textType -> Action Value
+--   into text using the given 'PandocWriter'.
+--   Returns a 'Value' wherein the rendered text is set to the "content" key and
+--   any metadata is set to its respective key in the 'Value'
+loadUsing :: PandocReader textType
+          -> PandocWriter
+          -> textType
+          -> Action Value
 loadUsing reader writer text = do
   (pdoc, meta) <- makePandocReader reader text
   outText      <- unPandocM $ writer pdoc
   let withContent = meta & _Object . at "content" ?~ String outText
   return withContent
 
 -- | Like 'loadUsing' but allows also deserializes the 'Value' into any object
--- which implements 'FromJSON'.  Failure to deserialize will fail the Shake
--- build.
-loadUsing'
-  :: (FromJSON a)
-  => PandocReader textType
-  -> PandocWriter
-  -> textType
-  -> Action a
-loadUsing' reader writer text = loadUsing reader writer text >>= convert
-
--- | Attempt to convert between two JSON serializable objects (or 'Value's).
--- Failure to deserialize fails the Shake build.
-convert :: (FromJSON a, ToJSON a, FromJSON b) => a -> Action b
-convert a = case fromJSON (toJSON a) of
-  Success r   -> pure r
-  Error   err -> fail $ "json conversion error:" ++ err
-
--- | Flatten a Pandoc 'Meta' into a well-structured JSON object, rendering Pandoc
--- text objects into plain strings along the way.
-flattenMeta :: Meta -> Value
-flattenMeta (Meta meta) = toJSON $ fmap go meta
- where
-  go :: MetaValue -> Value
-  go (MetaMap     m) = toJSON $ fmap go m
-  go (MetaList    m) = toJSONList $ fmap go m
-  go (MetaBool    m) = toJSON m
-  go (MetaString  m) = toJSON m
-  go (MetaInlines m) = toJSON $ stringify m
-  go (MetaBlocks  m) = toJSON $ stringify m
+--   which implements 'FromJSON'.  Failure to deserialize will fail the Shake
+--   build.
+loadUsing' :: (FromJSON a)
+           => PandocReader textType
+           -> PandocWriter
+           -> textType
+           -> Action a
+loadUsing' reader writer text =
+  loadUsing reader writer text >>= convert
+
+-- | Generalized Entity Loader, wrapper in
+--   own function to respolve EntityFilePath properly
+loadEntity :: FromJSON b
+           => ReaderOptions      -- ^ Pandoc reader options to specify extensions or other functionality
+           -> WriterOptions      -- ^ Pandoc writer options to modify output
+           -> EntityFilePath a   -- ^ Path to entity ready for conversion, usually text file
+           -> Action b
+loadEntity rops wops (EntityFilePath path) =
+  subloadEntity rops wops path
+
+-- | Helper functiona for generalized `loadEntity`
+--   provides conversion from source to output file format
```

Comment:
The docs don't make any mention of the fact that this uses markdown, packs to html, or depends on `url` and `srcPath`; 

Do you think this provides enough **general** value to be in the lib rather than at user-level?

Reply:
I moved code to `subloadEntity` to cover cases when not only markdown used. Some bloggers use `.rst` or `.tex`. I thought it might be cool to extend `loadEntity` function further and, in future, add argument which takes in function like `fnConvertEntityMyWay`(subloadEntity now). I don't want to do this complication before initial functionality merged. Can be easily moved back to `loadEntity` or generalized further by pushing conversion function to user space and adding default markdown converter here.

Reply:
I think this is starting to get into the area of being a bit more opinionated than I'd like slick to be; I want it to have nice defaults, but this adds constraints on:

* The file-type blogs are written in
* The output type we're building to
* The meta format/fields we use

Making all those assumptions to save someone 10 lines of user code seems like too much of a cost; I'd prefer this, `loadEntity`, and `EntityFilePath` be removed; you're more than welcome to write a `slick-blog-helpers` package or whatever with any helpers you want, since they don't depend on any internals it should be easy enough to extract. Let's keep the core lean and adaptable 😄

Reply:
Hm, the use case for `loadEntity` is to reduce code duplication, when library user doesn't have special rules for loader but cache types are different. Instead of multiple `loadPost`, `loadTeam`, etc. with almost identical code one can write: 
```haskell
  aboutCache         <- jsonCache' loadEntity
  teamCache          <- jsonCache' loadEntity
  postsCache         <- jsonCache' loadEntity
  postsDraftCache    <- jsonCache' loadEntity
  projectsCache      <- jsonCache' loadEntity
```
seems rather elegant to me. Example site have only posts, so I left customized version `loadPosts` for showcase purposes on customization. But if you think that weakens codebase we can explore options to create Slick goodies package like `slick-extra`.

Reply:
I see what you mean, but it still makes all the assumptions I mentioned, which isn't clear at all from the title.

I'd be open to keeping this if it was called `loadHTMLEntityFromMarkdownFile` and the `subloadEntity` was just not exported since it doesn't provide any benefit over the cached version. It would also need to specify clearly in the docs that it saves the `url` and `srcPath` at those locations, and that it.

It's a wordy name, but at least it tries to explain what it's doing. I agree this is a pretty common use-case, but unless it's explained well it's just a lot of "magic"